### ISSUE TITLE
mechanic request:  outfitter price scaling

### ISSUE THREAD

**Issue by virmitio** [endless-sky/endless-sky#1823]
I think it would add a great deal to the game to have the same outfit be potentially valued differently at different outfitters.  As an example, atomic engines might be more highly valued in the South or on the Rim (say a factor of 1.2x?) as compared to in the Deep.  Deprecation of outfits would still apply, but it would mean that there might become a market for transporting outfits for sale instead of traditional commodity cargo.  It would also make the sale of plundered outfits potentially more lucrative if one sells farther from where the outfit was acquired.

Another potential use would be to permit some outfits to be more readily available for sale (or available for sale at all) at a cost far higher than the currently assigned value of the good.  For example, in the future it might be possible to visit some civilization far in the galactic east which has jump drives for sale as standard equipment, but cost perhaps 10M each, while still selling in human space (where they can't be purchased at all) for their current "value" of 1M (before deprecation).

A possible implementation would be to have the "outfitter" entry on planets accept an optional second parameter as a scaling factor/multiplier for those outfits on that planet (defaulting to 1 to avoid compatibility issues).  Presumably this would only accept positive floats, but I'm sure someone else could come up with a practical meaning for negative values that I'm not thinking of right now.

**Comment by Amazinite** [comment]
Related #928

**Comment by virmitio** [comment]
As a clarification: While I can see the apparent relation to #928, what I'm really asking for is something that can be placed in the hands of the content/plugin makers instead of a static internal value adjuster.

**Comment by KiLEdEnNis** [comment]
still, same base concept

**Comment by jafdy** [comment]
The most obvious way of setting this up ether is a lot of work (price of every outfit defined to every outfitter) or it will run into some/most/all of the problems me and @LocalGod79 discussed in #928.

**Comment by virmitio** [comment]
jafdy, are you meaning a lot of work for the developer, or for the content creators?  I can't speak for the developer, as I haven't forked the code to see what would be necessary to implement it as I described yet.  For content creators, my proposed usage would look like the below adjustment in `map.txt`:

```
planet Longjump
	...
	outfitter "Basic Outfits"
	outfitter "Delta V Basics" .95
	outfitter "Ammo South"
	...

planet "New China"
	...
	outfitter "Delta V Basics" 1.05
	outfitter "Kraz Basics"
	...
```

The above is not intended as a _wise_ example of locations/outfits to use the scaling with, merely a syntax example.  Again, all of this assumes that standard outfit deprecation still applies.  In the above example, this would mostly mean that if one loots small plasma drives from pirates in the Sarin system, it would be more lucrative to sell them on Longjump than on New China, but buying those same drives new would be more affordable on New China.

A possible use for negative values to potentially avoid concerns of price abuse:
factor < -1.0 : Scale purchase cost by positive of value given, sell price unchanged.
-1.0 < factor < 0 : Scale sell price by positive of value given, purchase price unchanged.
factor in [-1, 0, 1] : prices unchanged from default 

These would give the option of locations where one may be able to purchase rare goods at a steep premium, but the merchant will only buy them back at "galactic market" rate.  This makes me think of trying to haggle for a particularly interesting piece of equipment with a trade-centric race, only to discover that I paid the cost of a house for their equivalent of a toaster oven.

**Comment by jafdy** [comment]
So this is a feature request for a plugin.
By a lot of work I mean someone will have to go through the map file to put the values in. Then they need to check that every value can not become an exploit and otherwise generally balance it. There must be over 100 planets to go through. In short it would be a dreadful waste of time for MZ and no sane modder would do it.

**Comment by virmitio** [comment]
I'm thinking I'll need to fork and make a pull request to mate up with this request to illustrate intent.  Adding the functionality as described would not require any changes whatsoever to the existing map data.  Any entry with no scaling factor provided would be handled exactly as it presently is.

I honestly have no expectation that anything existing would be changed to retroactively use it.  I just feel it would be beneficial to have this or a similarly accessible method available to allow differential pricing of identical outfits in differing market locales.

**Comment by Elyssaen** [comment]
Am I right that you could use a feature like this to give the player a discount?

e.g.

```
event "free world love you now"
	shipyard "Southbound Advanced" 0.8
```

...though whether setting up a black market Dreadnought market is a valid strategy for a Free Worlds Council member is another question. Alternatively, an early Dreadnought or Kestrel could cost more to buy than it's worth.

Seems like this could be potentially useful as a feature, though maybe not a high priority unless anyone's got something awesome to add to the game that's begging for it.

**Comment by KiLEdEnNis** [comment]
> Alternatively, an early Dreadnought or Kestrel could cost more to buy than it's worth.

speaking about the Kestrel, after the player is able to get it, do you ever see a non-MZ legit NPC Kestrel? no

**Comment by DingusShingleton** [comment]
Did someone say marauder kestrel?

**Comment by KiLEdEnNis** [comment]
1st we need actual Kestrels when they are unlocked
(and if I'd make my 1st pull request, it'd require adding an event when one of the missions is completed etc.
not hard to do in my files, but idk how to pr with several edited files

**Comment by endless-sky** [comment]
It's already hard enough, when outfitting a ship, to fly around to wherever all the outfits you want are sold. If you also needed to hunt for the best deal on each outfit, that could get annoying really fast.

**Comment by virmitio** [comment]
@Elyssaen While I hadn't initially though of this being used for ships as well, I see no reason that your example shouldn't or wouldn't also be viable.

@endless-sky I'm  not advocating for making sweeping (or really any) changes to the existing economy within the current systems.  I honestly expect that this, if implemented, would only show up in mods or in late-game (read as "Somewhere/sometime after the current end of missions").  To provide a more complete reason for the request, I'll give a bit more background into how I'd like to use it:

I'm currently working on a mod which would introduce new locations and multiple new starfaring races.  Each would have unique, though occasionally similar, outfits available.  The following situation is likely to arise multiple times throughout the mod's missions: 
- Race A and Race B are nominally at war, so free trade is no longer open between their respective systems.
- Race A's unique outfits would be worth considerably more for sale or purchase at Race B's ports during this time, and vice versa .
- At the conclusion of the war (if any), markets would at some point return to price parity in their respective outfitters.

I was also considering the possibility of having Jump Drives available for sale among one or more of these far-flung races, but only at a significant premium over the "market rate" one could sell them for in human space.  The initial cost in my head would put them around 40M each, but to accomplish that at the moment, I would either need to:
- Make a new "Jump Drive"-like outfit, which would have its own pricing, availability, and may require additional code changes to make it function in-game; or 
- Alter the price of the existing "Jump Drive" outfit in the mod's config files.  This would greatly alter the money-making dynamic around the time of the Pug war, as fleecing a few extra jump drives would suddenly net enough cash to purchase a new fleet. 

I'm open to other means of accomplishing the results I've described above.  The solution I proposed was just the simplest I could come up with that would be transparent to the existing config files and game experience if unused.

**Comment by LocalGod79** [comment]
from a modding perspective, this is already possible; there's nothing stopping you from creating two complete sets of outfits, and missions similar to the Unfettered Hai Jump Drive missions that offer a higher rate (payment) for "foreign" outfit sales (removal).

**Comment by virmitio** [comment]
@LocalGod79 You're correct that I could use either of those methods, however each carries significant side effects.

The first option of duplicating outfits means that my outfit **X** from race A still has a static value across all systems when I try to sell it (allowing for deprecation of value, of course).  This means there's still no way for me to value the same outfit higher during wartime on race B planets unless I make it more valuable for all planets.

The second option of having missions for the improved pricing would mean cluttering a player's mission list as well as bypassing outfit value deprecation.  ("Sure this **X** is the genuine article from race A, but it's worn down to the bearings all over.  You expect me to pay the full bounty on it?")  Still accessing the deprecated values from mission configs would also require additional code work, so I'm not sure this would be any more effective a solution or any less work to implement.

I'm not trying to denounce either of these options.  Just explaining why I rejected them as viable solutions for my desired result, thus leading to this feature request.

**Comment by LocalGod79** [comment]
i'm sorry, i seem to have confused you with my perhaps overly concise description.  there were not two options - it was one encompassing concept.

for race A, create a set of outfits.
for race B, create a set of outfits.
for race A, create a set of missions that pay more than market value for race B's outfits. (for which, distinct outfits from race B are required)
for race B, create a set of missions that pay more than market value for race A's outfits. (for which, distinct outfits from race A are required)

both sets of missions should require a condition that is applied to the player when the war begins, and is removed when the war ends.

sure it bypasses depreciation, but they're not buying them at a higher rate because they want to use them - they want to find better defence against them, or reverse engineer them for other purposes - surely both civilian factions don't value their opponent's weaponry more than their own (or somebody's failing in the propaganda war)...make them relatively cheap (for their progress at the time they'd encounter these aliens), so the player can't abuse non-depreciation...

**Comment by virmitio** [comment]
That does clear that suggestion up quite a bit, @LocalGod79.  My apologies for misunderstanding.

While that would address a single instance of the precise situation I described (or multiple instances of same situation if I want to duplicate a great deal of mission content), it only addresses this most recently described line item case.  It also, by your own admission, still ignores outfit deprecation; a feature which I find highly desirable to still apply specifically in order to provide a more diversified and interesting multi-species economy.  

As was pointed out by @Amazinite at the start of this thread, there are strong similarities between this request and #928, in that my initial request would effectively permit an outfit transport economy to be created.  As I've already expressed that I feel this has no present value in the existing data and serves only to introduce nuance in mods or for well-established players, I will be very understanding if further discussion or consideration needs to wait on a pull request.

**Comment by LocalGod79** [comment]
my own time has become even more limited lately, so suggestions that make work for our only coding developer (who also happens to create nearly all of the story content) are understandably low-priority i.m.h.o. - that being said a scaling attribute for outfit pricing seems as close as possible to what i had originally hoped for when i raised #928.

it should be noted that the discussion in #928 seemed to have a strong impact in making ammo have mass; implementation meant smaller mass missile outfits, with hoppers that would reduce the overall mass of the ship as ammo was expended, and obviously the cargo space impact.

with all of that said, until the "always available outfitter" is implemented, this might be the solution i hoped for when i proposed #928, and all that's required is a pull request with mapped sales, and that when the "always available outfitter" *is* implemented, that the sale value be 0 - essentially abandoned on a outfitter-less planet.  this may give rise to "i left my stuff on this planet, why isn't it still there X days later?", but that might be a coding enhancement for 1.0.1...

tbh, the issue that M.Z. raised with *this* issue would be easily solved by implementing my Ursa Polaris plugin (please pardon the advert), but it wouldn't feel right until the community had come up with the Republic and Syndicate storylines...(and so **there** is our proverbial roadblock)

**Comment by Wrzlprnft** [comment]
Source code change suggestion:I am not sure if local wanted to say something like this, or if it is what the O.P. initially ment, but what about some "thruster-flate scale" treatment for outfitters? Content creators can specify a number behind a outfits name in the outfitter, which represents the price change, and the game just defaults to 1 if no number is given?
Said number would be multiplied with the price.
Since we already have a that mechanic(though, for something different), it might be less work, and not immediatly impact the base game.

**Comment by Elyssaen** [comment]
Though I absolutely agree with endless-sky that doing this everywhere would actually just frustrate the player, I _do_ see interesting uses. Even in the content I'm working on at the moment. I also personally like the idea that you could use this to 'properly' implement this:

```
outfit "Intrusion Countermeasures"
	category "Hand to Hand"
	thumbnail "outfit/unknown"
	"capture defense" 60
	"unplunderable" 1
	description "Quarg ships are outfitted with special systems for repelling boarders. Because 
no one wants to be caught by the Quarg with stolen technology, these systems are pretty much
worthless on the open market."
```

because the Heliarch _might_ feel differently, and it's not clear why this only applies to Countermeasures.

I have a few other priorities in my life right now, but I am _very_ happy to try some code enhancement that isn't on his or anyone else's immediate to-do list. I first need to make a pull request for #1826 and I'd like to make a few weapon/outfit attributes for my own stuff, but then I could try tackling this. But only if MZ would be okay with this being made an option in the game.

Aside: sorry to divert the course of the thread, @LocalGod79, but I'd like to hear about state of community work on the Republic and Syndicate storylines. Who's doing what, etc. Do you know? If we want to keep this thread on track, I'd love to hear from you outside: my Skype's the same as this user name, and my email's rather easy to guess (I use Gmail).

**Comment by Wrzlprnft** [comment]
Local's the one coordinating republic story efforts, and as far as i know, it is at least layed out in a solid way, though it needs a lot of writing. The syndicate story effort was last tackled by a steamuser called messymix, though they seem to be stuck where we (as in local and his team) were before with the rep story: get a solid order into the brainstorm.

**Comment by virmitio** [comment]
@Wrzlprnft That usage was part of the proposal, as described in my second reply on this thread.  The possible usage was further expanded upon in later thread replies.  I'm presently working on a possible implementation of this for @endless-sky to review for use/modification/discarding.

@Elyssaen I think cases such as that might fall under a different code change.  Perhaps one where outfits at the present outfitter with a calculated (after scaling/deprecation) sell price of 0 cannot be sold?

**Comment by Elyssaen** [comment]
I don't think it's a problem that you can sell the 'unsellable' Intrusion Countermeasures, I just assume you threw them in an landfill or jettisoned them when you set off again. Actually making it mechanically unsellable seems a fiddly frustration if you then have to jettison it manually.

What I'd thinking is implement it with support for:

```
outfit "Intrusion Countermeasures"
	category "Hand to Hand"
	"cost" 50000
        "scale" 0
```

```
planet HeliarchWorld
    ...
    outfitter "Quarg" 1 "sell"
```

**Comment by virmitio** [comment]
That's an interesting thought.  To include an alternate "default" scale in the outfit definition.  For compatibility it would, again, have to assume `1` if not present, but that shouldn't be a particularly arduous task.  I'll see if such can be included in my present work.  Final decisions will, of course, be up to @endless-sky.

**Comment by Wrzlprnft** [comment]
I just realized that my phone cut off part of the code tag since it is not allowed to format it, which caused me to post a absolutely redundant post....

Back to the topic: i recommend just making a as simple as possible, focused on a single feature/version based pull request, if you can code it.
I would not put it in the outfit code, since it would need a certain failsafe, in case an outfitter or government does not exist. If you put it at an outfitter, those failsafe mechanics are already in place. Additionally, it could be modified with events, e.g. a price change, easier.

Though go with whatever you feel like :)

**Comment by Elyssaen** [comment]
Is that a 'no' to doing this as I was suggesting? I've had this on my code to-do list, though a low priority.

**Comment by endless-sky** [comment]
It seems like the uses would be rather limited, and if it were used too widely (forcing the player to "shop around" for the best prices) it'd be very annoying.
